src/Obelisk/Frontend/GoogleAnalytics.hs

```diff
@@ -1,54 +1,250 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
 {-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -Wall #-} -- -Werror #-}
 
 module Obelisk.Frontend.GoogleAnalytics
   ( googleAnalyticsFromConfig
   , googleAnalytics
   , trackingIdPath
   , getTrackingId
+  , Analytics(..)
+  , GoogleAnalyticsT(..)
+  , GtagJSCall(..)
+  , gaClickEvent
+  , gaOutboundClickEvent
+  , runGoogleAnalyticsT
   ) where
 
-import Control.Monad (void)
+import Control.Monad
+import Control.Monad.Exception
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Primitive
+import Control.Monad.Ref
+import Control.Monad.State.Strict
 import Data.Text (Text)
-import GHCJS.DOM.Types (liftJSM)
 import Language.Javascript.JSaddle.Evaluate (eval)
 import Obelisk.Configs
 import Obelisk.Route.Frontend
 import Reflex.Dom.Core
 import qualified Data.Text as T
+import qualified Data.Aeson as Aeson
+import Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+import Language.Javascript.JSaddle
+import Reflex.Host.Class
+
+import qualified Data.HashMap.Strict as HMap
+
+class Monad m => Analytics t w m | m -> w where
```

Comment:
Why parameterize this on `w`? As is, it's just `EventWriter` with a different name.

Reply:
Not quite;  it is specifying `Seq` as the "transport" container.   The idea is that this was ending up in an `Analytics` module,  then particular binding to `gtag.js` would be in a `Analytics.Google` module.   (By the way,  there are like, at least 3 or 4 officially supported javascript client libraries provided by Google... I haven't spent any time exploring the differences.  So really, there probably shouldn't be a `Analytics.Google` module, but rather something like an `Analytics.Google.GtagJS` module)

Ryan mentioned wanting to have some sort of application-specific way of specifying events;  then you could write a mapping from those events onto one or more different analytics services/backends,  which is probably a good idea.